README.md
# srcds-log-parser

![build & release](https://github.com/blastorg/srcds-log-parser/actions/workflows/release.yaml/badge.svg)
![package downloads](https://img.shields.io/npm/v/%40blastorg/srcds-log-parser)

A log parsing tool, that can be used for parsing logs sent from CS2 game server via an HTTP endpoint.

Big shout out to the original author of this library [negezor](https://github.com/negezor), who created this awesome library which strongly types log events coming from the game server.

## Reasoning behind the fork of this library

The reason why we have created this fork of the library, is because we saw that the library hasn't been actively worked on for about 2 years.

We have changed this library up a bit after Valve, the creators of Counter Strike, released their version 2 of the game (Counter Strike 2). They changed up some formats of how they are outputting the logs from the game server, and this library is up-to-date on that format.

## What's the difference between this library and the original one?

We have chosen to remove the log receiver because we didn't see a use case for it, the main part of this library should be as the name implies `srcds-log-parser`

We have added a few things which update the parser to be able to parse the new CS2 logs. The other part which has changed is the format the steam id is returned in. In the logs, the steam id includes the steam universe, which the steam account is connected to. To give developers and users who uses this library a better experience, we have chosen to convert the steam id coming from the logs to steam id 64 (decimal) format, which can be stored as a `BigInt`, because of it's size.

## Requirements

In order to install and use this package, you need to have Node.js installed (at least `v18`).

Because this library is using regex for all of it's parsing, it does not need any other dependencies.

# Installation

```bash
$ npm install @blastorg/srcds-log-parser
```

## Example usage

```ts
import { parse } from '@blastorg/srcds-log-parser';

// Log coming from CS2 game server
const log = '10/20/2020 - 10:30:50: "AttackerName<93><[U:1:230970467]><CT>" [698 2222 -69] killed "VictimName<94><[U:1:230970467]><TERRORIST>" [1303 2143 64] with "hkp2000" (throughsmoke headshot)'

const parsedLog = parse(log);

console.log(parsedLog);
{
  type: "killed",
  receivedAt: Date("10/20/2020 10:30:50");
  payload: {
    attacker: {
      kind: "player",
      entityId: 93,
      steamId: "76561198191236195", // Steam ID 64 (BigInt)
      name: "AttackerName",
      position: [698, 2222, -69],
      team: {
        id: 3,
        name: "COUNTER_TERRORISTS",
      },
    },
    victim: {
      kind: "player",
      entityId: 94,
      steamId: "76561198191236195", // Steam ID 64 (BigInt)
      name: "VictimName",
      position: [1303, 2143, 64],
      team: {
        id: 2,
        name: "TERRORISTS",
      },
    },
    weaponName: "hkp2000",
    modifiers: ["throughsmoke", "headshot"],
  },
}
```


CHANGELOG.md
# [2.1.0](https://github.com/blastorg/srcds-log-parser/compare/v2.0.3...v2.1.0) (2023-12-01)


### Bug Fixes

* **left_buyzone_with:** changes payload to match entity triggered event type ([98fee4c](https://github.com/blastorg/srcds-log-parser/commit/98fee4cac83b5fa58a98fc1932976b0e9338eb06))


### Features

* **parsers:** adds parser for parsing left-buyzone-with ([001e89f](https://github.com/blastorg/srcds-log-parser/commit/001e89fc2382fa4a519ee2da46e8b2e7c07b8c7a))

## [2.0.3](https://github.com/blastorg/srcds-log-parser/compare/v2.0.2...v2.0.3) (2023-11-03)


### Bug Fixes

* **ci:** fixes versioning on github repository for version 3.0.4 ([50a342d](https://github.com/blastorg/srcds-log-parser/commit/50a342d8fcd8e5408c15ef6d757476efb07e63f7))
* **ci:** uses plugin to get semantic new version ([8b863eb](https://github.com/blastorg/srcds-log-parser/commit/8b863eb90b721c28b0df0a84b01f4be7a614b0aa))
* **release:** removes release section from package.json ([564ce07](https://github.com/blastorg/srcds-log-parser/commit/564ce07e96d88770a515b351340975fe84c5856c))

## [2.0.3](https://github.com/blastorg/srcds-log-parser/compare/v2.0.2...v2.0.3) (2023-11-03)


### Bug Fixes

* **ci:** uses plugin to get semantic new version ([8b863eb](https://github.com/blastorg/srcds-log-parser/commit/8b863eb90b721c28b0df0a84b01f4be7a614b0aa))
* **release:** removes release section from package.json ([564ce07](https://github.com/blastorg/srcds-log-parser/commit/564ce07e96d88770a515b351340975fe84c5856c))

## [3.0.4](https://github.com/blastorg/srcds-log-parser/compare/v3.0.3...v3.0.4) (2023-11-03)


### Bug Fixes

* **ci:** uses plugin to get semantic new version ([e1ad82a](https://github.com/blastorg/srcds-log-parser/commit/e1ad82a67dbeb8ee8cd9d30493a235e8871d4a7d))

## [3.0.3](https://github.com/blastorg/srcds-log-parser/compare/v3.0.2...v3.0.3) (2023-11-03)


### Bug Fixes

* **ci:** removes .output from semantic release ([f403735](https://github.com/blastorg/srcds-log-parser/commit/f4037351ef259afd5c76e199e9897798a08130dd))

## [3.0.2](https://github.com/blastorg/srcds-log-parser/compare/v3.0.1...v3.0.2) (2023-11-03)


### Bug Fixes

* **ci:** removes needs from publish package workflow ([cc940e2](https://github.com/blastorg/srcds-log-parser/commit/cc940e22354f3589e16a19a4f64a76c90fa9eb75))

## [3.0.1](https://github.com/blastorg/srcds-log-parser/compare/v3.0.0...v3.0.1) (2023-11-03)


### Bug Fixes

* **ci:** moves publish package steps to release workflow ([4c69c81](https://github.com/blastorg/srcds-log-parser/commit/4c69c81d72b5ab456ed63362676d983f4f09d1d9))

# [3.0.0](https://github.com/blastorg/srcds-log-parser/compare/v2.0.10...v3.0.0) (2023-11-03)


### Bug Fixes

* **breaking:** testing if breaking change bumps correct version ([7468db6](https://github.com/blastorg/srcds-log-parser/commit/7468db6cf10e969f272d4807152b6a7507cbcbec))


### BREAKING CHANGES

* **breaking:** tests if bumping version works

## [2.0.10](https://github.com/blastorg/srcds-log-parser/compare/v2.0.9...v2.0.10) (2023-11-03)


### Bug Fixes

* **ci:** removes dollar and curly brackets from shell command ([b1c221d](https://github.com/blastorg/srcds-log-parser/commit/b1c221d7a914ed7fd0b762b1de2209d0dcc08834))

## [2.0.9](https://github.com/blastorg/srcds-log-parser/compare/v2.0.8...v2.0.9) (2023-11-03)


### Bug Fixes

* **ci:** adds git describe inside of npm version command ([36a9b35](https://github.com/blastorg/srcds-log-parser/commit/36a9b35ffd3783585697db84cf13730fab4fbcf3))
* **ci:** removes double curly brackets from npm version command ([6a54cb0](https://github.com/blastorg/srcds-log-parser/commit/6a54cb0bc289dad1c1924edc07629fe54339bbb5))

## [2.0.8](https://github.com/blastorg/srcds-log-parser/compare/v2.0.7...v2.0.8) (2023-11-03)


### Bug Fixes

* **ci:** adds fetch-tags to checkout action ([c07d482](https://github.com/blastorg/srcds-log-parser/commit/c07d482c9e329ebebeafb50dee80d7e08007c3e8))

## [2.0.7](https://github.com/blastorg/srcds-log-parser/compare/v2.0.6...v2.0.7) (2023-11-03)


### Bug Fixes

* **ci:** adds npm version parameters ([da19636](https://github.com/blastorg/srcds-log-parser/commit/da19636a126f743cac26403dfc707e7eba085058))

## [2.0.6](https://github.com/blastorg/srcds-log-parser/compare/v2.0.5...v2.0.6) (2023-11-03)


### Bug Fixes

* **ci:** adds parameters for npm version command ([9ee9f23](https://github.com/blastorg/srcds-log-parser/commit/9ee9f23d1d01c2cb9352310d34103c912b7aaeb0))

## [2.0.5](https://github.com/blastorg/srcds-log-parser/compare/v2.0.4...v2.0.5) (2023-11-03)


### Bug Fixes

* **ci:** adds from-git parameter to npm version command ([7d5b2eb](https://github.com/blastorg/srcds-log-parser/commit/7d5b2ebf46654523d28d2796e26ff3c176960d43))

## [2.0.4](https://github.com/blastorg/srcds-log-parser/compare/v2.0.3...v2.0.4) (2023-11-03)


### Bug Fixes

* **ci:** fixes release workflow ([d769b3b](https://github.com/blastorg/srcds-log-parser/commit/d769b3b20ffd90684a4ea595fc93271ad6270ba8))
* **ci:** fixes releasing workflow to deploy to both registries ([d0d2d41](https://github.com/blastorg/srcds-log-parser/commit/d0d2d41716a78baae344868c5de75d5018422a83))


src/parsers/killed.ts
import { IBaseEvent, defineParser } from "./parser";
import { concatPattern } from "../helpers";
import { Entity, entityRe, vectorRe, parseEntity, parseVector } from "../entities";

export type KilledEventPayload = {
  attacker: Entity;

  victim: Entity;

  weaponName: string;
  modifiers: ("headshot" | "penetrated" | "throughsmoke" | "noscope")[];
};

export type KilledEvent = IBaseEvent<"killed", KilledEventPayload>;

// eslint-disable-next-line max-len
// "AttackerName<93><STEAM_1:0:12345><CT>" [698 2222 -69] killed "VictimName<94><STEAM_1:0:12345><TERRORIST>" [1303 2143 64] with "hkp2000" (throughsmoke headshot)
export const killedParser = defineParser<KilledEvent>({
  type: "killed",

  patterns: [
    concatPattern`^(?<attacker>${entityRe}) \\[(?<attackerPosition>${vectorRe})\\] killed (?<victim>${entityRe}) \\[(?<victimPosition>${vectorRe})\\] with "(?<weaponName>[^"]+)"(?: \\((?<modifiers>[^\\)]+)\\))?$`,
  ],

  parse({
    attacker,
    attackerPosition,

    victim,
    victimPosition,

    weaponName,
    modifiers,
  }) {
    return {
      attacker: {
        ...parseEntity(attacker),

        position: parseVector(attackerPosition),
      },
      victim: {
        ...parseEntity(victim),

        position: parseVector(victimPosition),
      },

      weaponName,

      modifiers: (modifiers?.split(" ") ?? []) as KilledEventPayload["modifiers"],
    };
  },
});
